packages/legend-studio/src/stores/sidebar-state/WorkspaceSyncState.ts

```diff
@@ -0,0 +1,584 @@
+  }
+}
+
+export class WorkspaceSyncState {
+  editorStore: EditorStore;
+  sdlcState: EditorSDLCState;
+
+  isSyncingWorkspace = false;
+  incomingRevisions: Revision[] = [];
+  workspaceSyncConflictResolutionState: WorkspaceSyncConflictResolutionState;
+
+  constructor(editorStore: EditorStore, sdlcState: EditorSDLCState) {
+    makeAutoObservable(this, {
+      editorStore: false,
+      sdlcState: false,
+      fetchIncomingRevisions: flow,
+      setIncomingRevisions: action,
+      syncWorkspace: flow,
+      resetConflictState: action,
+      forceSync: flow,
+      synWithHeadAndLoadChanges: flow,
+    });
+
+    this.editorStore = editorStore;
+    this.sdlcState = sdlcState;
+    this.workspaceSyncConflictResolutionState =
+      new WorkspaceSyncConflictResolutionState(editorStore, sdlcState);
+  }
+
+  resetConflictState(): void {
+    this.workspaceSyncConflictResolutionState.teardown();
+    this.workspaceSyncConflictResolutionState =
+      new WorkspaceSyncConflictResolutionState(
+        this.editorStore,
+        this.sdlcState,
+      );
+  }
+
+  setIncomingRevisions(revisions: Revision[]): void {
+    this.incomingRevisions = revisions;
+  }
+
+  *fetchIncomingRevisions(): GeneratorFn<void> {
+    try {
+      assertTrue(this.sdlcState.isWorkspaceOutOfSync);
+      const revisions = (
+        (yield this.editorStore.sdlcServerClient.getRevisions(
+          this.sdlcState.activeProject.projectId,
+          this.sdlcState.activeWorkspace,
+          this.sdlcState.activeRevision.committedAt,
+          this.sdlcState.activeRemoteWorkspaceRevision.committedAt,
+        )) as PlainObject<Revision>[]
+      ).map((e) => Revision.serialization.fromJson(e));
+      this.setIncomingRevisions(
+        revisions.filter((r) => r.id !== this.sdlcState.activeRevision.id),
+      );
+    } catch (error) {
+      this.setIncomingRevisions([]);
+      assertErrorThrown(error);
+    }
+  }
+
+  *syncWorkspace(): GeneratorFn<void> {
+    try {
+      assertTrue(this.sdlcState.isWorkspaceOutOfSync);
+      this.editorStore.setBlockingAlert({
+        message: `Syncing workspace...`,
+        showLoading: true,
+      });
+      this.isSyncingWorkspace = true;
+      const changes =
+        this.editorStore.changeDetectionState.workspaceLocalLatestRevisionState
+          .changes;
+      let conflicts: EntityChangeConflict[] = [];
+      if (changes.length) {
+        // Base revision <-> Local
+        yield flowResult(
+          this.editorStore.changeDetectionState.computeLocalChanges(),
+        );
+        yield flowResult(
+          this.editorStore.changeDetectionState.computeAggregatedWorkspaceRemoteChanges(),
+        );
+        conflicts =
+          this.editorStore.changeDetectionState.potentialWorkspaceSyncConflicts;
+      }
+      if (conflicts.length) {
+        this.editorStore.setBlockingAlert(undefined);
+        this.editorStore.setActionAltertInfo({
+          message: 'Conflicts found while syncing workspace',
```

Comment:
Please change all `syncing` teminology to `pushing changes` or `pull changes` accordingly like we discussed last time. I don't have any immediate suggestion but like we said last time, let's pretty much eradicate the usage of the word `sync` because we don't `sync`, we have decomposed it to `push` and `pull`